### DISCUSSION TITLE
SQL validation errors - how to make more specific in Production without risk

### DISCUSSION THREAD

**Discussion by LSNicholls** [serenity-is/Serenity#6657] [Help]
Hi folks,

There is a lot of good information in #6409 on this subject... 

I would like to provide better messages to the client than "An error occurred on your request" without opening up to full Development messages where inappropriate.  

For example,  I have some uniqueness constraints that can't rely on the Unique attribute for a nice message because the uniqueness requirement is not a single column, it's "unique within a scope".  A typical use case would be a list of survey questions where the sequence # for each question should be unique within the survey ID value (which is a FK).  

If I run this in Development, of course I get the full database error from the constraint.  In Release or Production, I get "An error occurred on your request" -- and I would like to intercept this and provide something better. 

I would be open to doing this either as a validation step, ie before save, and I see there is some way of creating behaviors  which would be called in ValidateRequest.

Meanwhile, I do have these constraints, as well as a calculated column with a concatenated value, available on the SQL side.  I tried exposing the concatenated key value in the row and giving it the Unique attribute, which doesn't work, for reasons that are obvious in hindsight:  that value doesn't exist yet on an uncreated or unsaved row.  

I think it *might* work to AddBehavior() to these concatenated key values. I might even be able to do this with one behavior object that would work across all the different tables that have this requirement.

But ... I think it would be easier to create a behavior of type ISaveExceptionBehavior, or maybe just edit <project>.HumanizeSqlExceptionBehavior. and <project>.SqlExceptionHelper, plus Texts.cs, in this case.

I am going to try the latter, but was wondering if anybody had experience with this that they wanted to share?

**Comment by edwardch** [comment]
Unique constraints aren't in the SqlExceptionHelper by default. Mine is a slightly older version (still waiting for the right time to do upgrades), but in there is a `HandleSaveException` method. By default, it calls `HandleSavePrimaryKeyException`, then I added a `HandleSaveUniqueIndexException` method to it, which looks something like this:

```cs
public static void HandleSaveUniqueIndexException(Exception e, ITextLocalizer localizer)
{
    PrimaryKeyExceptionInfo pk;
    if (SqlExceptionHelper.IsUniqueIndexException(e, out pk))
        throw new ValidationError(String.Format(Texts.Validation.SaveUniqueIndexError.ToString(localizer), pk.TableName, pk.Value));
}

public static bool IsUniqueIndexException(Exception e, out PrimaryKeyExceptionInfo pk)
{
    // The SQL Error number for a unique index violation is 2601
    // sample: Cannot insert duplicate key row in object 'dbo.Customers' with unique index 'UQ_Customers_IdNumber'. The duplicate key value is (12345).
    // Cannot insert duplicate key row in object '%.*ls' with unique index '%.*ls'. The duplicate key value is %ls.

    var sql = e as SqlException;
    if (sql != null && sql.Errors.Count > 0 && sql.Errors[0].Number == 2601)
    {
        var msg = sql.Errors[0].Message;
        pk = new PrimaryKeyExceptionInfo();
        pk.TableName = "???";

        var s1 = "in object '";
        var idx = msg.IndexOf(s1);
        if (idx >= 0)
        {
            idx += s1.Length;
            var s2 = "' with unique";
            var idx2 = msg.IndexOf(s2, idx + 1);
            if (idx2 >= 0)
            {
                pk.TableName = msg.Substring(idx, idx2 - idx);
                if (pk.TableName.StartsWith("dbo."))
                    pk.TableName = pk.TableName.Substring("dbo.".Length);
            }
        }

        var vs1 = "duplicate key value is (";
        var vidx = msg.IndexOf(vs1);
        if (vidx >= idx)
        {
            vidx += vs1.Length;
            var vs2 = ")";
            var vidx2 = msg.IndexOf(vs2, vidx + 1);
            if (vidx2 >= 0)
            {
                pk.Value = msg.Substring(vidx, vidx2 - vidx);
            }
        }

        return true;
    }

    pk = null;
    return false;
}
```

I haven't looked at the newer version to see whether it's in this helper or in behaviours, but maybe this helps.

**Comment by LSNicholls** [comment]
Thank you, Edward, that is almost exactly what I did yesterday!  Except my parsing code is different because I don't really want them seeing the duplicate key value, as in most of the cases where the native Unique attribute would be handling the whole thing nicely the value would be something like (5,16).  This would be not only uninformative but strange to the user since I hide the key values for most entities anyway.

Also I did create a separate UniqueKeyExceptionInfo class, not sure I really needed to.

For completeness, we both also needed to add the Texts.Validation.SaveUniqueIndexError entry.

For anybody who is interested in whether this is different in the newer versions, I think from what Edward is saying that **HumanizeSqlExceptionBehavior** is a new class, and this is where you **put SqlExceptionHelper.HandleSaveUniqueKeyException(...)** now.

However.... in reading #6409, I got the feeling that Victor might have been recommending a different, newer approach using behaviors.

![image](https://user-images.githubusercontent.com/88156108/212486862-e3567762-4492-40f3-b817-1ecce3937689.png)

I do see the ISaveExceptionBehavior interface, and I *think* (although I am not sure) that I would use the AddBehavior() attribute on the relevant calculated key columns, but I am not sure how to do this.

I had another thought, that maybe this is what setting CheckBeforeSave to false on the Unique attribute would be for.  But I think that the error would still occur on the Save, so I don't see how this would work in practice.

So, Edward, I'm not going to mark your reply as the answer just yet, although I really appreciate the confirmation (and it does work) <g>.

Hoping there will be other ideas that people can share here?

**Comment by VictorTomaili** [comment]
You can use the [HumanizeSqlExceptionBehavior](https://github.com/serenity-is/serene/blob/master/src/Serene.Web/Modules/Common/Helpers/HumanizeSqlExceptionBehavior.cs#L8) as an example. You can modify it or create a new one. [Here is where primary key exception](https://github.com/serenity-is/serene/blob/611d4da9b753c134757496f0fe401aa8ffb32309/src/Serene.Web/Modules/Common/Helpers/SqlExceptionHelper.cs#L27) message is throws. 

I try to mean in the message, throwing exception after validate the data is a nice pattern. And he can validate the primary server by a behavior.

**Comment by LSNicholls** [comment]
Thank you Victor.  I guess I didn't explain my question very well.  I don't have a problem understanding where the exception is thrown, and I easily found HumanizeSqlExceptionBehavior.  But I don't understand how HumanizeSqlExceptionBehavior is registered so that the framework knows to call it.  

I found it easy to edit HumanizeSqlExceptionBehavior, in other words, but I couldn't see how to add a separate behavior into the collection that would be used, rather than editing this one.

And thank you for explaining your previous message.  If that user wanted to do what you were suggesting, where would he register that behavior? 

I'm afraid this question is probably stupid from your POV but I really haven't figured it out.

>L<

**Reply by VictorTomaili** [reply]
The behaviors work by checking each row for conditions. You can write any condition [in this line](https://github.com/serenity-is/serene/blob/master/src/Serene.Web/Modules/Common/Helpers/HumanizeSqlExceptionBehavior.cs#L12), like implemented interface requirements, attribute, some fields with exact name or an attribute on a field etc. If this method returns true, then this behavior is executed. By default, this method returns true to execute on all rows.

**Comment by LSNicholls** [comment]
Victor, are you saying that it just has to be compiled into the project and implement the appropriate interfaces?  Does it need to be in any particular namespace to work? 

Again, apologies for not understanding this.  And thank you again.

**Reply by VictorTomaili** [reply]
Here is document https://serenity.is/docs/tutorials/multi_tenancy/using_serenity_service_behaviors

**Reply by LSNicholls** [reply]
I am so sorry I missed this part of the docs.  Thanks again.
